test/e2e/scheduling/predicates.go

```diff
@@ -724,6 +727,231 @@ var _ = SIGDescribe("SchedulerPredicates [Serial]", func() {
+			err = wait.Poll(pollInterval, wait.ForeverTestTimeout, deploymentReplicasScheduled(ctx, cs, ns, deploymentPodLabels))
+			framework.ExpectNoError(err)
+
+			// new replicas shouldn't be affected by existing replicas.
+			framework.Logf("Make sure all Pods in updated deployment %q is scheduled on the same Node", deploymentName)
+			nodenames, err = deploymentReplicasNodeNames(ctx, cs, ns, deploymentPodLabels)
+			framework.ExpectNoError(err)
+			nodename2 := nodenames[0]
+			for _, n := range nodenames {
+				framework.ExpectEqual(nodename2, n)
+			}
+
+			// The second deployment should be scheduled on a Node different from the first deployment's Pods.
+			framework.ExpectNotEqual(nodename2, nodename)
+		})
+
+		ginkgo.It("PodAntiAffinity: Pods from the same replicaset are taken into account by matchLabelKeys during the deployment's rolling upgrade [Feature:MatchLabelKeysInPodAffinity] [alpha]", func(ctx context.Context) {
+			// It's a real-world usecase of matchLabelKeys.
+			// Create a deployment, run rolling upgrade, and see new replicas are spread over all Nodes.
+			// The replica num will be set to be the same as the number of Node.
+			// So, if the matchLabelKeys feature is broken, no new Pod can be scheduled during the rolling upgrade and this test fails.
+
+			e2eskipper.SkipUnlessNodeCountIsAtLeast(3)
+			numNode := framework.TestContext.CloudConfig.NumNodes
+			ginkgo.By("make the nodes have balanced resource usage ratio")
+			err := createBalancedPodForNodes(ctx, f, cs, ns, nodeList.Items, podRequestedResource, 0.5)
+			framework.ExpectNoError(err)
+
+			k := v1.LabelHostname
+			// All of Pods hate each other and get scheduled in a different node.
+			affinity := &v1.Affinity{
+				PodAntiAffinity: &v1.PodAntiAffinity{
+					RequiredDuringSchedulingIgnoredDuringExecution: []v1.PodAffinityTerm{
+						{
+							LabelSelector: &metav1.LabelSelector{
+								MatchExpressions: []metav1.LabelSelectorRequirement{
+									{
+										Key:      "security",
+										Operator: metav1.LabelSelectorOpIn,
+										Values:   []string{"S1"},
+									},
+								},
+							},
+							TopologyKey:    k,
+							Namespaces:     []string{ns},
+							MatchLabelKeys: []string{"pod-template-hash"},
+						},
+					},
+				},
+			}
+
+			intstrPtr := func(i int) *intstr.IntOrString {
+				num := intstr.FromInt(i)
+				return &num
+			}
+			deploymentPodLabels := map[string]string{"test": "matchLabelKey1", "security": "S1"}
+			deploymentName := "test-matchlabelkey-deployment"
+			deploymentReplicas := int32(numNode)
+			deploymentImage := imageutils.GetPauseImageName()
+			deploymentStrategyType := appsv1.RollingUpdateDeploymentStrategyType
+			newDeployment := e2edeployment.NewDeployment(deploymentName, deploymentReplicas, deploymentPodLabels, "pause", deploymentImage, deploymentStrategyType)
+			newDeployment.Spec.Strategy.RollingUpdate = &appsv1.RollingUpdateDeployment{
+				MaxUnavailable: intstrPtr(0),
```

Comment:
nit: just use intstr.IntOrString directly
```suggestion
				MaxUnavailable: &intstr.IntOrString{IntVal: 0},
```